server/src/main/java/com/broadcom/lsp/cobol/service/delegates/references/ElementOccurrences.java

```diff
@@ -0,0 +1,104 @@
+
+import com.broadcom.lsp.cobol.core.semantics.outline.RangeUtils;
+import com.broadcom.lsp.cobol.service.CobolDocumentModel;
+import com.broadcom.lsp.cobol.service.delegates.validations.AnalysisResult;
+import lombok.NonNull;
+import lombok.Value;
+import org.eclipse.lsp4j.*;
+
+import javax.annotation.Nullable;
+import java.util.*;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * This occurrences provider resolves the requests for the semantic elements based on its positions.
+ */
+public class ElementOccurrences implements Occurrences {
+    @Override
+    public @NonNull List<Location> findDefinitions(@Nullable CobolDocumentModel document, @NonNull TextDocumentPositionParams position) {
+        if (document == null) return Collections.emptyList();
+        return findElementByPosition(document, position).definitions;
+    }
+
+    @Override
+    public @NonNull List<Location> findReferences(@Nullable CobolDocumentModel document, @NonNull TextDocumentPositionParams position, @NonNull ReferenceContext context) {
+        if (document == null) return Collections.emptyList();
+        Element element = findElementByPosition(document, position);
+        List<Location> references = element.usages;
+        if (context.isIncludeDeclaration()) {
+            references = new ArrayList<>(references);
+            references.addAll(element.definitions);
+        }
+        return references;
+    }
+
+    @Override
+    public @NonNull List<DocumentHighlight> findHighlights(@Nullable CobolDocumentModel document, @NonNull TextDocumentPositionParams position) {
+        return findReferences(document, position, new ReferenceContext(true)).stream()
+                .filter(byUri(position))
+                .map(toDocumentHighlight())
+                .collect(Collectors.toList());
+    }
+
+    private static Element findElementByPosition(CobolDocumentModel document, TextDocumentPositionParams position) {
+        AnalysisResult result = document.getAnalysisResult();
+        return findElementByPosition(result.getVariableDefinitions(), result.getVariableUsages(), position)
+                .or(() -> findElementByPosition(result.getParagraphDefinitions(), result.getParagraphUsages(), position))
+                .or(() -> findElementByPosition(result.getSectionDefinitions(), result.getSectionUsages(), position))
+                .or(() -> findElementByPosition(result.getConstantDefinitions(), result.getConstantUsages(), position))
+                .or(() -> findElementByPosition(result.getCopybookDefinitions(), result.getCopybookUsages(), position))
+                .or(() -> findElementByPosition(result.getSubroutineDefinitions(), result.getSubroutineUsages(), position))
+                .orElseGet(() -> new Element(List.of(), List.of()));
+    }
+
+    @Value
+    private static class Element {
```

Comment:
I don't think that this can be reused in any other place. It contains just definitions and usages, to be able to use it somewhere we must add at least a name.
Well, the name is not needed here, and the class is pointless without a name. I would like to keep it private.